File: swift/Sources/threads/Thread.swift
import Foundation
import Yams

// Status constants
let terminalStatuses = ["resolved", "superseded", "deferred"]
let activeStatuses = ["idea", "planning", "active", "blocked", "paused"]
let allStatuses = activeStatuses + terminalStatuses

// FrontmatterRaw is the raw parsed YAML with optional fields
struct FrontmatterRaw: Codable {
    var id: String?
    var name: String?
    var desc: String?
    var status: String?
}

// Frontmatter represents the YAML frontmatter of a thread
struct Frontmatter: Codable {
    var id: String
    var name: String
    var desc: String
    var status: String

    init(id: String = "", name: String = "", desc: String = "", status: String = "idea") {
        self.id = id
        self.name = name
        self.desc = desc
        self.status = status
    }

    init(from raw: FrontmatterRaw) {
        self.id = raw.id ?? ""
        self.name = raw.name ?? ""
        self.desc = raw.desc ?? ""
        self.status = raw.status ?? ""
    }
}

// Thread represents a parsed thread file
class Thread {
    var path: String         // absolute file path
    var frontmatter: Frontmatter
    var content: String      // full file content
    var bodyStart: Int       // character offset where body starts (after frontmatter)

    init(path: String, frontmatter: Frontmatter, content: String, bodyStart: Int) {
        self.path = path
        self.frontmatter = frontmatter
        self.content = content
        self.bodyStart = bodyStart
    }

    // Parse reads and parses a thread file
    static func parse(path: String) throws -> Thread {
        let content = try String(contentsOfFile: path, encoding: .utf8)

        let thread = Thread(
            path: path,
            frontmatter: Frontmatter(),
            content: content,
            bodyStart: 0
        )

        try thread.parseFrontmatter()

        // Extract ID from filename if not in frontmatter
        if thread.frontmatter.id.isEmpty {
            thread.frontmatter.id = extractIDFromPath(path)
        }

        return thread
    }

    // parseFrontmatter extracts and parses YAML frontmatter
    func parseFrontmatter() throws {
        guard content.hasPrefix("---\n") else {
            throw ThreadError.missingFrontmatterDelimiter
        }

        // Find closing delimiter
        let searchStart = content.index(content.startIndex, offsetBy: 4)
        guard let endRange = content.range(of: "\n---", range: searchStart..<content.endIndex) else {
            throw ThreadError.unclosedFrontmatter
        }

        let yamlContent = String(content[searchStart..<endRange.lowerBound])
        bodyStart = content.distance(from: content.startIndex, to: endRange.upperBound) + 1 // skip past \n---\n

        // Parse YAML - use FrontmatterRaw to handle optional fields
        let decoder = YAMLDecoder()
        do {
            let raw = try decoder.decode(FrontmatterRaw.self, from: yamlContent)
            frontmatter = Frontmatter(from: raw)
        } catch {
            throw ThreadError.yamlParseError(error.localizedDescription)
        }
    }

    // ID returns the thread ID
    var id: String { frontmatter.id }

    // Name returns the thread name/title
    var name: String { frontmatter.name }

    // Status returns the thread status
    var status: String { frontmatter.status }

    // baseStatus returns status without reason suffix
    var baseStatus: String {
        Thread.baseStatus(frontmatter.status)
    }

    // baseStatus strips reason suffix from status
    static func baseStatus(_ status: String) -> String {
        if let idx = status.range(of: " (") {
            return String(status[..<idx.lowerBound])
        }
        return status
    }

    // isTerminal returns true if the status is terminal
    static func isTerminal(_ status: String) -> Bool {
        let base = baseStatus(status)
        return terminalStatuses.contains(base)
    }

    // isValidStatus returns true if the status is valid
    static func isValidStatus(_ status: String) -> Bool {
        let base = baseStatus(status)
        return allStatuses.contains(base)
    }

    // body returns the content after frontmatter
    var body: String {
        guard bodyStart < content.count else { return "" }
        let idx = content.index(content.startIndex, offsetBy: bodyStart)
        return String(content[idx...])
    }

    // setFrontmatterField updates a frontmatter field and rewrites content
    func setFrontmatterField(_ field: String, _ value: String) throws {
        switch field {
        case "id":
            frontmatter.id = value
        case "name":
            frontmatter.name = value
        case "desc":
            frontmatter.desc = value
        case "status":
            frontmatter.status = value
        default:
            throw ThreadError.unknownField(field)
        }

        try rebuildContent()
    }

    // rebuildContent reconstructs file content from frontmatter and body
    func rebuildContent() throws {
        var sb = "---\n"

        // Marshal frontmatter
        let encoder = YAMLEncoder()
        let yaml = try encoder.encode(frontmatter)
        sb += yaml
        sb += "---\n"

        // Preserve body
        if bodyStart < content.count {
            let idx = content.index(content.startIndex, offsetBy: bodyStart)
            sb += String(content[idx...])
        }

        content = sb
    }

    // write saves the thread to disk
    func write() throws {
        try content.write(toFile: path, atomically: true, encoding: .utf8)
    }

    // relPath returns the path relative to workspace
    func relPath(_ ws: String) -> String {
        if path.hasPrefix(ws) {
            var rel = String(path.dropFirst(ws.count))
            if rel.hasPrefix("/") {
                rel = String(rel.dropFirst())
            }
            return rel
        }
        return path
    }
}

// extractIDFromPath extracts the 6-char hex ID from a filename
func extractIDFromPath(_ path: String) -> String {
    let filename = (path as NSString).lastPathComponent
    let base = (filename as NSString).deletingPathExtension

    // Check for ID prefix pattern: abc123-slug
    let pattern = #"^([0-9a-f]{6})-"#
    if let regex = try? NSRegularExpression(pattern: pattern),
       let match = regex.firstMatch(in: base, range: NSRange(base.startIndex..., in: base)),
       let range = Range(match.range(at: 1), in: base) {
        return String(base[range])
    }
    return ""
}

// extractNameFromPath extracts the human-readable name from a filename
func extractNameFromPath(_ path: String) -> String {
    let filename = (path as NSString).lastPathComponent
    let base = (filename as NSString).deletingPathExtension

    // Check for ID prefix pattern: abc123-slug
    let pattern = #"^[0-9a-f]{6}-"#
    if let regex = try? NSRegularExpression(pattern: pattern),
       regex.firstMatch(in: base, range: NSRange(base.startIndex..., in: base)) != nil {
        // Skip the ID prefix (7 chars: 6 hex + dash)
        return String(base.dropFirst(7))
    }
    return base
}

enum ThreadError: Error, LocalizedError {
    case missingFrontmatterDelimiter
    case unclosedFrontmatter
    case yamlParseError(String)
    case unknownField(String)

    var errorDescription: String? {
        switch self {
        case .missingFrontmatterDelimiter:
            return "missing frontmatter delimiter"
        case .unclosedFrontmatter:
            return "unclosed frontmatter"
        case .yamlParseError(let msg):
            return "YAML parse error: \(msg)"
        case .unknownField(let field):
            return "unknown field: \(field)"
        }
    }
}

File: go/internal/thread/thread.go
package thread

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"gopkg.in/yaml.v3"
)

// Status constants
var (
	TerminalStatuses = []string{"resolved", "superseded", "deferred", "reject"}
	ActiveStatuses   = []string{"idea", "planning", "active", "blocked", "paused"}
	AllStatuses      = append(ActiveStatuses, TerminalStatuses...)
)

// Frontmatter represents the YAML frontmatter of a thread
type Frontmatter struct {
	ID     string `yaml:"id"`
	Name   string `yaml:"name"`
	Desc   string `yaml:"desc"`
	Status string `yaml:"status"`
}

// Thread represents a parsed thread file
type Thread struct {
	Path        string      // absolute file path
	Frontmatter Frontmatter // parsed YAML frontmatter
	Content     string      // full file content
	BodyStart   int         // byte offset where body starts (after frontmatter)
}

// Note represents an item in the Notes section
type Note struct {
	Hash string
	Text string
}

// TodoItem represents a checkbox item in the Todo section
type TodoItem struct {
	Hash    string
	Text    string
	Checked bool
}

// idPrefixRe matches ID-prefixed filenames like "abc123-slug-name.md"
var idPrefixRe = regexp.MustCompile(`^([0-9a-f]{6})-`)

// Parse reads and parses a thread file
func Parse(path string) (*Thread, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}

	t := &Thread{
		Path:    path,
		Content: string(content),
	}

	if err := t.parseFrontmatter(); err != nil {
		return nil, fmt.Errorf("parsing frontmatter: %w", err)
	}

	// Extract ID from filename if not in frontmatter
	if t.Frontmatter.ID == "" {
		t.Frontmatter.ID = ExtractIDFromPath(path)
	}

	return t, nil
}

// parseFrontmatter extracts and parses YAML frontmatter
func (t *Thread) parseFrontmatter() error {
	content := t.Content

	if !strings.HasPrefix(content, "---\n") {
		return fmt.Errorf("missing frontmatter delimiter")
	}

	// Find closing delimiter
	end := strings.Index(content[4:], "\n---")
	if end == -1 {
		return fmt.Errorf("unclosed frontmatter")
	}

	yamlContent := content[4 : 4+end]
	t.BodyStart = 4 + end + 4 // skip opening ---, yaml, closing ---, and newline

	if err := yaml.Unmarshal([]byte(yamlContent), &t.Frontmatter); err != nil {
		return err
	}

	return nil
}

// ExtractIDFromPath extracts the 6-char hex ID from a filename
func ExtractIDFromPath(path string) string {
	filename := filepath.Base(path)
	filename = strings.TrimSuffix(filename, ".md")

	if m := idPrefixRe.FindStringSubmatch(filename); len(m) > 1 {
		return m[1]
	}
	return ""
}

// ExtractNameFromPath extracts the human-readable name from a filename
func ExtractNameFromPath(path string) string {
	filename := filepath.Base(path)
	filename = strings.TrimSuffix(filename, ".md")

	if m := idPrefixRe.FindStringSubmatch(filename); len(m) > 1 {
		return filename[7:] // skip "abc123-"
	}
	return filename
}

// ID returns the thread ID
func (t *Thread) ID() string {
	return t.Frontmatter.ID
}

// Name returns the thread name/title
func (t *Thread) Name() string {
	return t.Frontmatter.Name
}

// Status returns the thread status
func (t *Thread) Status() string {
	return t.Frontmatter.Status
}

// BaseStatus returns status without reason suffix (e.g., "blocked (waiting)" -> "blocked")
func (t *Thread) BaseStatus() string {
	return BaseStatus(t.Frontmatter.Status)
}

// BaseStatus strips reason suffix from status
func BaseStatus(status string) string {
	if idx := strings.Index(status, " ("); idx != -1 {
		return status[:idx]
	}
	return status
}

// IsTerminal returns true if the status is a terminal status
func IsTerminal(status string) bool {
	base := BaseStatus(status)
	for _, s := range TerminalStatuses {
		if s == base {
			return true
		}
	}
	return false
}

// IsValidStatus returns true if the status is valid
func IsValidStatus(status string) bool {
	base := BaseStatus(status)
	for _, s := range AllStatuses {
		if s == base {
			return true
		}
	}
	return false
}

// Body returns the content after frontmatter
func (t *Thread) Body() string {
	if t.BodyStart >= len(t.Content) {
		return ""
	}
	return t.Content[t.BodyStart:]
}

// SetFrontmatterField updates a frontmatter field and rewrites content
func (t *Thread) SetFrontmatterField(field, value string) error {
	switch field {
	case "id":
		t.Frontmatter.ID = value
	case "name":
		t.Frontmatter.Name = value
	case "desc":
		t.Frontmatter.Desc = value
	case "status":
		t.Frontmatter.Status = value
	default:
		return fmt.Errorf("unknown field: %s", field)
	}

	return t.rebuildContent()
}

// rebuildContent reconstructs file content from frontmatter and body
func (t *Thread) rebuildContent() error {
	var sb strings.Builder

	sb.WriteString("---\n")

	// Marshal frontmatter
	fm, err := yaml.Marshal(&t.Frontmatter)
	if err != nil {
		return err
	}
	sb.Write(fm)
	sb.WriteString("---\n")

	// Preserve body (BodyStart points to content after closing ---\n)
	if t.BodyStart < len(t.Content) {
		sb.WriteString(t.Content[t.BodyStart:])
	}

	t.Content = sb.String()
	return nil
}

// Write saves the thread to disk
func (t *Thread) Write() error {
	return os.WriteFile(t.Path, []byte(t.Content), 0644)
}

// RelPath returns the path relative to workspace
func (t *Thread) RelPath(ws string) string {
	rel, err := filepath.Rel(ws, t.Path)
	if err != nil {
		return t.Path
	}
	return rel
}
